client/lib/streamframer/framer.go

```diff
@@ -0,0 +1,310 @@
+package framer
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+	"time"
+)
+
+var (
+	// HeartbeatStreamFrame is the StreamFrame to send as a heartbeat, avoiding
+	// creating many instances of the empty StreamFrame
+	HeartbeatStreamFrame = &StreamFrame{}
+)
+
+// StreamFrame is used to frame data of a file when streaming
+type StreamFrame struct {
+	// Offset is the offset the data was read from
+	Offset int64 `json:",omitempty"`
+
+	// Data is the read data
+	Data []byte `json:",omitempty"`
+
+	// File is the file that the data was read from
+	File string `json:",omitempty"`
+
+	// FileEvent is the last file event that occurred that could cause the
+	// streams position to change or end
+	FileEvent string `json:",omitempty"`
+}
+
+// IsHeartbeat returns if the frame is a heartbeat frame
+func (s *StreamFrame) IsHeartbeat() bool {
+	return s.Offset == 0 && len(s.Data) == 0 && s.File == "" && s.FileEvent == ""
+}
+
+func (s *StreamFrame) Clear() {
+	s.Offset = 0
+	s.Data = nil
+	s.File = ""
+	s.FileEvent = ""
+}
+
+func (s *StreamFrame) IsCleared() bool {
+	if s.Offset != 0 {
+		return false
+	} else if s.Data != nil {
+		return false
+	} else if s.File != "" {
+		return false
+	} else if s.FileEvent != "" {
+		return false
+	} else {
+		return true
+	}
+}
+
+// StreamFramer is used to buffer and send frames as well as heartbeat.
+type StreamFramer struct {
+	out chan<- *StreamFrame
+
+	frameSize int
+
+	heartbeat *time.Ticker
+	flusher   *time.Ticker
+
+	shutdown   bool
+	shutdownCh chan struct{}
+	exitCh     chan struct{}
+
+	// The mutex protects everything below
+	l sync.Mutex
+
+	// The current working frame
+	f    StreamFrame
+	data *bytes.Buffer
+
+	// Captures whether the framer is running and any error that occurred to
+	// cause it to stop.
+	running bool
+	err     error
+}
+
+// NewStreamFramer creates a new stream framer that will output StreamFrames to
+// the passed output channel.
+func NewStreamFramer(out chan<- *StreamFrame,
+	heartbeatRate, batchWindow time.Duration, frameSize int) *StreamFramer {
+
+	// Create the heartbeat and flush ticker
+	heartbeat := time.NewTicker(heartbeatRate)
+	flusher := time.NewTicker(batchWindow)
+
+	return &StreamFramer{
+		out:        out,
+		frameSize:  frameSize,
+		heartbeat:  heartbeat,
+		flusher:    flusher,
+		data:       bytes.NewBuffer(make([]byte, 0, 2*frameSize)),
+		shutdownCh: make(chan struct{}),
+		exitCh:     make(chan struct{}),
+	}
+}
+
+// Destroy is used to cleanup the StreamFramer and flush any pending frames
+func (s *StreamFramer) Destroy() {
+	s.l.Lock()
+
+	wasShutdown := s.shutdown
```

Comment:
Not sure if you need to keep track of wasShutdown, maybe consider:
```
if !s.shutdown(
  close(s.shutdownCh)
)
s.shutdown = true
```

Reply:
In this block you are right but I use it outside of the lock on line 124